### DISCUSSION TITLE
Lack of ServiceHost in CoreWCF

### DISCUSSION THREAD

**Discussion by mconnew** [CoreWCF/CoreWCF#522] [General]
Copied from comment by @bstordrup in #234:

System.ServiceModel namespace has a ServiceHost class that derives from System.ServiceModel.ServiceHostBase. I can see in the CoreWCF source that it implements a ServiceHostBase class, but for easing migration of existing projects from old WCF to CoreWCF, I think implementing a ServiceHost class would be good.

System.ServiceModel namespace has a generic ClientBase that is used for implementing a WCF service client. A similar class would be good to have in CoreWCF to ease migrating existing projects from old WCF to CoreWCF

**Comment by mconnew** [comment]
I deleted your comments on that issue as ServiceHost isn't ever going to be on the RoadMap.

In WCF the base class for hosting is actually ServiceHostBase. There are multiple hosting models, and they all derive from ServiceHostBase. In WCF, the various transports are encapsulated within it. So for example, for HTTP, WCF creates an instance of HttpListener internally that you never get to see. For NetTcp, it creates a listening socket that you never get to see.  

When CoreWCF was built on top of ASP.NET Core, that's no longer an option. If the same approach were done, you wouldn't be able to have a WebApi or MVC app on the same port as a WCF service as you would have no way to get to the embedded ASP.NET Core instance to add any more middleware. We would need to create api's to enable choosing and picking all the various HTTP hosts sys as HttpSys, Kestrel, IIS etc. CoreWCF would need to do a LOT more, and would need a lot more code so give you less flexibility. This would mean fewer features (as time would be spent on that), and a significant lag when new things are added to ASP.NET Core to expose some way of using them.  

CoreWCF is a library which extends ASP.NET Core, not encapsulate it. Just think for a moment if your starting point was ServiceHost, how would you inject something into DI? ServiceHost has the wrong flow of control so isn't ever going to be ported.

As to your second suggestion, CoreWCF is the service side of WCF. For client usage, there's https://github.com/dotnet/wcf which does contain ClientBase\<TChannel\>.

**Reply by Stolpe** [reply]
Thank you for clarifying that CoreWCF is for the service side of WCF, and also pointing to how to handle the client side. I found [this blog post](https://devblogs.microsoft.com/dotnet/upgrading-a-wcf-service-to-dotnet-6/)  which also says so in the subject. Somehow, I missed that detail, and it seems that I am not the only one. Your last phrase clearly stated what I needed to know for my use case, which is lifting an old project to a modern version of .Net.

**Reply by jimm98y** [reply]
@mconnew Well, I think it really depends upon the use case for this library. We have a product that's been in development for ~18 years and the problem now is how to migrate it from NET Framework to NET6. CoreWCF would have been quite useful if we were able to use it as a drop-in replacement, but without the ServiceHost it's not possible. In our case we are not looking for any new features and WCF is needed only for a backwards compatibility. Yes, the control flow is all backwards when compared to ASP.NET Core, but on the other hand, it means in order to upgrade to NET6 we'd have to rewrite all legacy components and change the flow in our app. So in our case, it'd be really helpful to have the drop-in replacement for ServiceHost, because the rewrite of everything that is currently using it would be too expensive.

**Reply by mconnew** [reply]
I think you're thinking this is a bigger change than it actually is. Where you previously would instantiate ServiceHost, now you use WebHost.CreateBuilder and go from there. If you have code which needs to access properties on ServiceHost, e.g. to modify the ServiceCredentials object, you can call `ConfigureServiceHostBase<TService>` which you provide a delegate to which takes a ServiceHostBase. It's mostly refactoring the startup code and you shouldn't need to rewrite anything significant.

**Reply by kachalkov** [reply]
We are in the same positiin, but mostly due to the fact that somehow DI can't return Transient registrations which we have for most of dependencies for our services. So, ServiceHostFactory works great for us, but we just getting that instance can't be created, what are the options in here?

**Comment by dretax** [comment]
@mconnew I've seen you guys have done the named pipe CoreWCF library. How would one do the implementation for this old .net framework code that im porting to standard? The web has to stay .framework for a while or so. The code is also not in the web project.
`    public class LogServiceHandler
    {
        ServiceHost _ServiceHost = null;
        static LogServiceHandler _Instance = null;

        LogServiceHandler()
        {
            string name = xyz.Instance.Configuration.ServerName.Trim().Replace(' ', '_');
            _ServiceHost = new ServiceHost(typeof(ServerLogService), new Uri("net.pipe://localhost/" + name + "Server"));
            _ServiceHost.AddServiceEndpoint(
                typeof(ILogService),
                new NetNamedPipeBinding(NetNamedPipeSecurityMode.None),
                "net.pipe://localhost/" + name + "Server/EndPoint"
                );
            _ServiceHost.Open();
        }`

**Comment by mconnew** [comment]
```c#
string name = xyz.Instance.Configuration.ServerName.Trim().Replace(' ', '_');
var builder = WebApplication.CreateBuilder();

builder.Services.AddServiceModelServices();
builder.WebHost.UseNetNamedPipe(options =>
    {
        options.Listen(new Uri("net.pipe://localhost/" + name + "Server"));
    });
var app = builder.Build();

app.UseServiceModel(serviceBuilder =>
{
    serviceBuilder.AddService<ServerLogService>();
    serviceBuilder.AddServiceEndpoint<ServerLogService, ILogService>(
        new NetNamedPipeBinding(NetNamedPipeSecurityMode.None),
        "net.pipe://localhost/" + name + "Server/Endpoint");
});

app.Start();
```

There's a small issue at the moment where asp.net core always starts an HTTP server listening which is an annoyance. I need to add support to use IHostBuilder and not use IWebHostBuilder in order to stop that happening.  

Also make sure to explicitly add a package reference to CoreWCF.NetFramingBase v1.5.2 as 1.5.1 and earlier (including the 1.5.0-preview1) versions have a vulnerability. The preview release of NetNamedPipe will work fine with the release build of NetFramingBase.

**Comment by hsedr** [comment]
Hey @mconnew, great work you all are doing with this project.

I have a question regarding the implementation of the `ServiceHost` class.
We are currently investigating alternatives to WCF at my company, and CoreWCF seems like a perfect fit to start modernizing the application. The only problem is that parts of the application cannot move to .NET (Core) because of dependencies like Windows WF. We are aware of the open source port of WF (CoreWF), but it is out of question as of now because it lacks key features like re-hosting.

This means that we are only able to move certain parts of the application to .NET (Core). But the other parts which will remain in .NET Framework are in need of the self hosting (ServiceHost) functionality of WCF. 

I understand that die `ServiceHost` implementation won't be added to the official repository. 
But how would one go about implementing this class themself? Is it at all worth looking into porting the `ServiceHost`? 
For us specifically not all functionalities are absolutely necessary, for example the partial configuration support through `App.config` is perfectly fine.
